drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+				    base_addr + CAN_XLNX_ICR_OFFSET);
+			LOG_ERR("%s read from RX FIFO caused an undeflow error",
+				dev->name);
+			break;
+		}
+	} while (isr & CAN_XLNX_IRQ_RX_NOT_EMPTY);
+}
+
+/**
+ * @brief TX FIFO transmission done handler function
+ * @internal
+ *
+ * Called from within the ISR if the TX OK interrupt bit is set in
+ * the Interrupt Status Register, this function executes each sent
+ * message's transmission done handler. This can either be the com-
+ * bination of a callback function pointer plus its parameter pointer,
+ * or a pointer to a semaphore on whose acquisition the can_xlnx_send
+ * function is currently blocking.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ */
+static void can_xlnx_handle_tx(const struct device *dev)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+	struct can_xlnx_dev_data *dev_data = dev->data;
+	struct can_xlnx_tx_done_handler *handler;
+
+	/* Clear the TX OK interrupt bit */
+	sys_write32(CAN_XLNX_IRQ_MESSAGE_TX, base_addr + CAN_XLNX_ICR_OFFSET);
+
+	/* Process all pending TX done handlers */
+	while (dev_data->tx_done_rd_idx != dev_data->tx_done_wr_idx) {
+		handler = &dev_data->tx_done_handlers[dev_data->tx_done_rd_idx];
+		dev_data->tx_done_rd_idx = (dev_data->tx_done_rd_idx + 1) %
+					   CAN_XLNX_RX_TX_FIFO_DEPTH;
+
+		if (handler->cb != NULL) {
+			handler->cb(dev, 0, handler->cb_data);
+		} else if (handler->sem != NULL) {
+			k_sem_give(handler->sem);
+		}
+
+		handler->cb = NULL;
+		handler->cb_data = NULL;
+		handler->sem = NULL;
+	}
+}
+
+/* Functions accessible via the driver's API */
+
+/**
+ * @brief Sets the current operational mode of the Xilinx CAN controller
+ *
+ * Sets the current controller instance's operational mode based on the
+ * mode parameter provided by the caller. The parameter's content is
+ * translated to the matching values to be written into the controller's
+ * Mode Select Register (MSR). Before the CAN bus state is being reset
+ * and interrupts are being re-enabled, it is assured that the specified
+ * operational mode has actually been reached.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ * @param mode Supported modes: CAN_MODE_NORMAL, CAN_MODE_LISTENONLY,
+ *             CAN_MODE_LOOPBACK. Unsupported: silent loopback mode
+ *             (CAN_MODE_LISTENONLY | CAN_MODE_LOOPBACK) and CAN_MODE_FD.
+ *
+ * @retval -ENOTSUP in case of invalid mode value, 0 otherwise.
+ */
+int can_xlnx_set_mode(const struct device *dev, can_mode_t mode)
+{
+	uintptr_t base_addr = DEVICE_MMIO_GET(dev);
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	struct can_xlnx_dev_data *dev_data = dev->data;
+#endif
+	uint32_t msr = 0x0;
+	uint32_t sr;
+	uint32_t ier;
+	uint8_t target_mode_reached = 0;
+
+	/*
+	 * Documentation of operational modes and transitions between
+	 * them: comp. Zynq-7000 TRM, chap. 18.2.1.
+	 */
+
+#ifdef CONFIG_CAN_FD_MODE
+	if (mode & CAN_MODE_FD) {
+		LOG_ERR("%s CAN FD is unsupported", dev->name);
+		return -ENOTSUP;
+	}
+#endif
```

Comment:
No need for the `#ifdef` here. This should always produce an error regardless of the Kconfig.